### ISSUE TITLE
Static linking on Windows broken

### ISSUE THREAD

**Issue by marcboeker** [marcboeker/go-duckdb#51]
When compiling the amalgamation file to build a static library the compiler complains about `definition is marked dllimport`. Others have this [problem too](https://github.com/duckdb/duckdb/discussions/2774#discussioncomment-3781570).

But when adding a `-DDUCKDB_BUILD_LIBRARY` the errors and warnings are gone and the resulting `libduckdb.a` is successfully created. But linking agains it throws the following errors (for more examples see [log.txt](https://github.com/marcboeker/go-duckdb/files/9792641/log.txt)):

```
C:\Users\dev\AppData\Local\Temp\go-link-2035893457\000007.o: in function `_cgo_37088fb8ef74_Cfunc_duckdb_bind_uint16':
/tmp/go-build/cgo-gcc-prolog:313: undefined reference to `__imp_duckdb_bind_uint16'
```

This is very weird as the Darwin and Linux builds succeed. The code is located under https://github.com/marcboeker/go-duckdb/tree/windows-builds

**Comment by candiduslynx** [comment]
Hi @marcboeker!
Can I suggest trying with https://github.com/ebitengine/purego instead of cgo?

**Comment by candiduslynx** [comment]
I also found the following issue that might be of use: https://github.com/golang/go/issues/51007#issuecomment-1369090750

Could you try reproducing with Go 1.20+?

**Comment by candiduslynx** [comment]
No, unfortunately. I was looking into the issue for https://github.com/cloudquery/cloudquery/issues/8854, but the builds take a couple of hours to complete with the results we already have (+ I believe, the code needs rebasing, as the DuckDB version in the branch is stale).

However, looking at the issue it seems that the thing to look at is this:
1. Check that the produced file is able to be linked statically
2. Verify that the static linking is used instead of the dynamic one (dynamic may be somehow enforced for Go based on the CGO being used, as the `__imp_` prefixes indicate an attempt to link against DLL instead of static linking).

**Comment by ZenLiuCN** [comment]
Hi, I just try to using static link under windows. 
1. compile a static lib from duckdb source (mingw)
2. link it with go-duckdb
3. run into the same trouble 
4. found inside duckdb.h
```c
//
// duckdb.h
//
//
//===----------------------------------------------------------------------===//

#pragma once

// duplicate of duckdb/main/winapi.hpp
#ifndef DUCKDB_API
#ifdef _WIN32
#if defined(DUCKDB_BUILD_LIBRARY) && !defined(DUCKDB_BUILD_LOADABLE_EXTENSION)
#define DUCKDB_API __declspec(dllexport)
#else
#define DUCKDB_API //__declspec(dllimport) // !! here 
#endif
```
5. comment out that party maks it no link error when compile.

**Comment by ayuhito** [comment]
Maybe this is related? https://github.com/duckdb/duckdb/issues/12571

**Comment by localhosted** [comment]
@ZenLiuCN Do you have more information on createing the static duckdb build and linking it to go? 

I haven't found any instructions on a static build in github.com/duckdb/duckdb/blob/main/Makefile

**Comment by taniabogatsch** [comment]
I just merged https://github.com/marcboeker/go-duckdb/pull/248, which bundles a pre-compiled static library for Windows. It will be part of the next go-duckdb release sometime next week.

**Comment by ZenLiuCN** [comment]
mingw compiled static libs may not supports all official extensions. suggest that should be noted in readme.

**Comment by taniabogatsch** [comment]
Fair point, I already added a small note to the `README`, but I'll extend it a bit. We're also looking into adding MingW distributions for our official extensions. We already build them for our `Rtools`, so we might just have to generalize those.

**Comment by ZenLiuCN** [comment]
Very good news for me. My last trying got blocked by mysql🤔
